### ISSUE TITLE
Deletion of empty cells workflow improvement

### ISSUE THREAD

**Issue by pawadski** [jupyterlab/jupyterlab#7713]
## Description

This is more of a feature request than it is a bug report. I was not able to find an already existing request for this.

I understand how to delete empty cells (ESC->D,D), but it is disappointing that I need to go through this every time I need to delete empty cells, instead of just pressing backspace while they are an active input. Why not make it so empty cells can be removed by pressing backspace?

## Reproduce

1. Create a notebook
2. Create some empty cells
3. Attempt to delete the empty cells by pressing backspace
4. Nothing happens

## Expected behavior

The empty cell is deleted

## Context

- Operating System and version: Windows 10 1903
- Browser and version: Brave 1.1.23 Chromium: 79.0.3945.88
- JupyterLab version: 1.1.4

**Comment by jasongrout** [comment]
I think this makes sense. It's similar to deleting a paragraph in a multi-paragraph document, and is analogous to adding a cell at the end when pressing shift-enter to execute the last cell.

**Comment by agbaum** [comment]
Is anyone else working on this? I'd like to take a stab at it.

Should it trigger in situations where the cell is empty, but just contains whitespace? And if so, should it do so in situations where the cell contains multiple empty lines, or just one line (which might have whitespace)? 

From an implementation standpoint, does it make sense to mimic the implementation of pressing up at the top of a cell to go the next cell up (emitting a signal from the codeMirrorEditor which gets handled by the notebook widget)?

**Comment by pawadski** [comment]
**From user perspective** I would appreciate only empty or whitespace blocks to be removed this way, _unless_ there is an undo for this operation.

**Comment by agbaum** [comment]
I agree that it doesn't make sense to do it for cells with actual text in them. I was more wondering if it should it happen for a cell that has whitespace in it.

**Comment by krassowski** [comment]
Just saw this merged. I am not necessarily happy with this being assigned to backspace. What bothers me is that this is is a destructive action if one tries to clean a cell and start writing its content anew. Un-doing the delete action requires two to three key strokes (Esc → shift + z, or Esc → b).

As an alternative, could we change it to shift-backspace?

**Comment by jasongrout** [comment]
> What bothers me is that this is is a destructive action if one tries to clean a cell and start writing its content anew. Un-doing the delete action requires two to key strokes (Esc → z, or Esc → b).
> 
> As an alternative, could we change it to shift-backspace (by analogy to shift-enter)?

Thanks for voicing an opinion about this. How do you imagine someone cleaning a cell? I imagine ctrl/cmd-A to select all the text, then pressing backspace once, or just starting to type the new content.

Shift-backspace would be a more conservative shortcut, but still would give some of the benefits if people knew about it.

**Comment by krassowski** [comment]
Indeed for a cell with a lot of code, it is <kbd>ctrl</kbd> + <kbd>A</kbd> followed by <kbd>backspace</kbd> or <kbd>ctrl</kbd> + <kbd>x</kbd>. But if this is just a few characters keeping <kbd>backspace</kbd> pressed continuously is often what people (including myself) do.

It is true that this is what you would expect in a text processor (removing a paragraph), but in a text, processor adding a new paragraph back is a single keystroke (<kbd>Enter</kbd>) but in notebook doing <kbd>shift</kbd> + <kbd>Enter</kbd> has side effects (re-running code which might be slow or otherwise not intended).

Again, no strong opinion, and I would not want to block this being resolved - just giving a user's opinion/worries. It might be that once it is released, I (and others) will just get used to it.

**Comment by pawadski** [comment]
Can you elaborate why removing a cell of nothing is destructive to your workflow? Do you happen to lose any data/code when removing an empty cell?

**Comment by krassowski** [comment]
No, it is not destructive to my workflow, but by definition, deletion is a destructive action.

An empty cell (as in "cell with no code or text") can still contain metadata (output/plot and/or tags) which would get removed. It is not very bad because the user can undo the deletion with <kbd>Esc</kbd> + <kbd>Z</kbd> - only potentially inconvenient. While I see that I am alone in my concerns here, there are many users who do not check GitHub periodically and I am certain that some other users can get surprised by this change (and not necessarily happy if they learn about this in practice). May I suggest that at the very least this should land high in the changelog notes?

**Comment by jasongrout** [comment]
> I might be just an outlier here.

I *really* appreciate you taking the time to explain your point. We're well aware that only a few people are participating here, but likely represent a lot more people that aren't tracking these issues.

So what do people think? Shift-Backspace is nearly as convenient, and wouldn't be as surprising to people used to the notebook never deleting cells in edit mode. I came at this from the Sage notebook, where we did have this behavior (IIRC), and I found it incredibly convenient.

@pawadski - what are your thoughts specifically, since you brought this up originally. Would shift-backspace be okay for your workflow?

**Comment by jasongrout** [comment]
Moving to 2.1 milestone so it stays at the top of our queue until we resolve this discussion, which may involve another PR to 2.1.

CC also @tgeorgeux and @ellisonbg, who have also paid a lot of attention to ux, but haven't commented here, and may not be aware of this discussion.

**Comment by pawadski** [comment]
I don't know but I might be an outlier too! It would be interesting to see how other people view this.

It seems consistent to have Shift+Backspace delete a cell, as you can use Shift+Enter to create cells (for example).

**Comment by jasongrout** [comment]
> I don't know but I might be an outlier too! It would be interesting to see how other people view this.

I think we're all guessing :).

I would support moving to shift-backspace as a more conservative option, and seeing if that is convenient enough.

**Comment by jasongrout** [comment]
We've now reverted the PR putting this in 2.1 pending further discussion of the keyboard shortcut: #8133. We are doing this because we are releasing 2.1 RC today.
